### ISSUE TITLE
Durable functions created hanging in pending state for long time, start up days later.

### ISSUE THREAD

**Issue by gpduck** [Azure/azure-functions-durable-extension#896]
### Description
Our durable functions are getting stuck in the pending state, sitting like that for days, and then magically starting.  This started after upgrading the nuget package versions, but I haven't been able to fix the issue by redeploying the old code, so I'm not sure if it's related.

Here are the package changes that were made:
Microsoft.Azure.DurableTask.Core 2.1.0 => 2.1.2
Microsoft.Azure.Management.Fluent 1.21.0 => 1.24.0
Microsoft.Azure.Management.ResourceManager.Fluent 1.21.0 => 1.24.0
Microsoft.Azure.Services.AppAuthentication 1.2.0-preview2 => 1.3.0
Microsoft.Azure.WebJobs.Extensions.DurableTask 1.8.0 => 1.8.3
Microsoft.EntityFrameworkCore.Design 2.2.4 => 2.2.6
Microsoft.EntityFrameworkCore.SqlServer 2.2.4 => 2.2.6
Microsoft.Graph 1.15.0 => 1.16.0
Microsoft.IdentityModel.Protocols.OpenIdConnect 5.4.0 => 5.5.0
Microsoft.IdentityModel.Tokens 5.4.0 => 5.5.0
Microsoft.NET.Sdk.Functions 1.0.24 => 1.0.29

The application consists of a UI front end that is calling an HTTP triggered function that starts a durable function.

I also have the same version of this code running in a separate function app that is not (currently) experiencing this issue.

### Expected behavior
Durable functions would start in a timely fashion.

### Actual behavior
Durable functions are sitting in the pending state for days before running

### Relevant source code snippets
```csharp
string instanceId = await orchestrationClient.StartNewAsync("SelfServiceRequestOrchestrator", ssRequest);
```

### Known workarounds
None

### App Details

- **Durable Functions extension version (e.g. v1.8.3)**: 1.8.3
- **Azure Functions runtime version (1.0 or 2.0)**: 2.0
- **Programming language used**: csharp

### If deployed to Azure
> We have access to a lot of telemetry that can help with investigations. Please provide as much of the following information as you can to help us investigate!

- **Timeframe issue observed**: (see below)
- **Function App name**: 
- **Function name(s)**: SelfServiceRequestOrchestrator
- **Azure region**: East US 2
- **Orchestration instance ID(s)**: (see below)
- **Azure storage account name**:

Request 1
- **Status**: Started successfully
- **InstanceID**: a4b34ae1d8be40309db77ca57e77b04f
- **Time**: 2019-08-14 17:28:04 (UTC)

Request 2
- **Status**: hung at pending state
- **InstanceID**: ef433fea4f074da49b00ce333725547c
- **Time**: 2019-08-14 17:29:14 (UTC)

Request 3
- **Status**: Started successfully
- **InstanceID**: ad0f0a15e54648db9fb16faa63ef4999
- **Time**: 2019-08-14 17:29:40 (UTC)

**Comment by marcin-synak** [comment]
I have a similar issue with one of my functions. I'm using the latest extension version (1.8.3) with Azure Functions v2. Multiple functions share the same storage account but use different hub names.

The function is a classic fan-in, fan-out scenario. It creates a bunch of tasks in a loop, waits until they all complete using Task.WhenAll and then should perform some activities to finish the job.

Roughly in 50% of cases, tasks are scheduled and completed (as can be seen in History table) but the additional activities are never scheduled. The whole orchestration gets stuck in "Running" state.

I have some tasks getting stuck after e.g. 20 minutes and some completing successfully after >2hrs (amout of work differs between tasks).

Account info:
- **Azure region:** West Europe
- **Function app name:** AdRakerCoreFunctionsRecommendations
- **Function name:** RecommendationsExport (orchestrator)
- **Azure storage account name:** adrakercorefunctions
- **Hub name:** RecommendationsHub

Example of execution stuck in "Running" state:
- **Orchestration instance ID (partition key):** 438901f3fdc442589741053a1d132e3e
- **Time:** 2019-08-16T01:13:17.298Z

Example of completed execution:
- **Orchestrator instance ID (partition key):** 2c7ccee859554819bd214ca9eb5bb9d9
- **Time:** 2019-08-16T01:22:04.666Z

**Comment by gpduck** [comment]
The hung task (ef433fea4f074da49b00ce333725547c) eventually started running at 2019-08-17 1:27:59 (UTC).  Let me know if there is any additional information I can provide to help identify what is going on.

**Comment by jtoniolo** [comment]
I'm seeing the same issue as @marcin-synak, but doesn't only get stuck on a fan-in/fan-out, although it has.

Where it stops seems completely random for us. We have a very long Orchestrator with many Activities and Sub-orchestrators and one fan-in/fan-out somewhere in the middle. Sometimes the main Orchestrator gets stuck in Running state, sometimes one of the Sub-orchestrators gets stuck.

App Insights doesn't appear to be logging any related exceptions.

Same issue with DF 1.8.3 and 1.7.1 (current version is 1.7.1, we rolled back to see if that would do the trick, it didn't)
Azure Functions v1

We first noticed this on 2019-08-16. Everything was running fine the previous weekend, from what I recall.

Account info:

Azure region: Canada East
Function app name: YMCAFunctionAppDEV
Function name: BulkCreate (orchestrator), BulkUpdate (orchestrator) or ProcessPaymentShares_ProcessBillDate (orchestrator)
Azure storage account name: ymcaappservicesstorage
Hub name: BDOTasksDEV
Example of execution stuck in "Running" state:

Orchestration instance ID (partition key): 009c86d95f704a54bfd9bf0270bdf133:422
Time: 2019-08-20T16:34:33.218Z

**Comment by graham-saunders** [comment]
I'm having the same issue and it's affecting our production environments for hundreds of tenants. It would be good if we could get confirmation that someone at Azure is looking into this

Account info:

Azure region: West US
Function app name: VirtuousIntJobsProd
Function name: ProcessStatisticsForOrganization
Azure storage account name: virtuousjobsprod
Hub name: InterminableJobsHubInstances
Example of execution stuck in "Running" state:

Orchestration instance ID (partition key): 1d216e87034b4571a3e633f148dc656b:235
Time: 2019-08-22T08:17:32.915Z

Example of completed execution:
Orchestrator instance ID (partition key): 1d216e87034b4571a3e633f148dc656b:233
Time: 2019-08-22T08:03:09.534Z

**Comment by ConnorMcMahon** [comment]
I am looking into this today. Thanks all for providing details, this will really help with the investigation.

**Comment by ConnorMcMahon** [comment]
@gpduck 

I took a look at your issue, which looking at the logs is quite a bit different than the issue @graham-saunders, @jtoniolo, and @marcin-synak are experiencing. I will open up a separate issue for what they are seeing in their applications.

As far as your application goes, it looks like for some reason we aren't detecting any messages in the control queue for your application until the app starts up several days later. When you have an instance stuck like this, can you check in your control queue directly and verify that there are messages to be processed in the queue? If there are, can you see if restarting your application unblocks the orchestrations. Reporting back on the status of this would be very helpful.

The fact that this issue and the issue the others are seeing is happening across all versions of Durable Functions makes me think something changed in the underlying platform (either Azure Functions runtime or the underlying App Service infrastructure). I am still investigating, but wanted to give a status update.

**Comment by ConnorMcMahon** [comment]
@gpduck

Also, I notice that on 8/12, your application was upgraded from Functions runtime host version 2.0.12590.0 to 2.0.12625.0. Could you [try pinning your function host to](https://docs.microsoft.com/en-us/azure/azure-functions/set-runtime-version) to 2.0.1259?

@graham-saunders, @jtoniolo, and @marcin-synak: that might be a good test to try for your issues as well.

**Comment by gpduck** [comment]
If the control queue is the durablefunctionshub-workitems queue in the storage account, I believe I checked using Azure Storage Explorer and I don't think anything was showing up there when it was stuck.  I just tried to reproduce the issue but it seems to be working fine now.  I'll be sure to check next time the issue comes up.  I also definitely tried restarting (and redeploying my code) when it was broken and neither ever fixed the stuck jobs.

I've tried pinning to 2.0.12590, but I'm getting an error:
```
Unsupported Runtime Version
Your custom runtime version is not supported. As a result 2.0.12641.0 runtime is being used. 
```

Is there any logs I have access to that would show the runtime version upgrades?  I'm curious about the prod version of this app that hasn't had any issues (it's currently on 2.0.12641)

**Comment by ConnorMcMahon** [comment]
@gpduck. 

That queue is actually the work item queue, where the Activity Functions are queued up. 

The control queues should be of the name `<taskhubname>-control-<num>`, where there are multiple different values for `<num>` starting at 00 (we default to 4, so you should see 00-03).

As far as versions go:

It looks like your app just hit 2.0.12641 as of about 35 min ago (2019-08-22 22:51:41.1691517 UTC to be exact).

If you have application insights set up, it should be under the column `sdkVersion` in the `traces` table.

Let's see if this issue is persisting with the new runtime version. I will see if there is another older version is still available in our system that we can pin to if that doesn't fix the issues.

**Comment by gpduck** [comment]
I'm pretty sure I checked all those as well and they were all empty, but I can't be sure at this point.

I've updated my pin version to 2.0.12641 so it will be stable and we can see how it goes.

**Comment by graham-saunders** [comment]
@ConnorMcMahon we're on ~1 still and the last release was 24 days ago so I'm hesitant to jump back a version when these issues only started in the past week. Can you let me know when you've opened the new issue for what we're seeing?

**Comment by cgillum** [comment]
@gpduck two important things I'm noticing in your case:

1. The message for **ef433fea4f074da49b00ce333725547c** was picked up by your app for the first time 3 days later, but with a dequeue count of **626**.
2. Your function app does not have a hubName configured in your host.json, so it's using the default task hub name of **DurableFunctionsHub**.

Most likely you have some other function app that is also listening for messages on the same storage account, and the two apps are competing with each other. That is the most common cause for the exact symptoms you're seeing. The other app is not running in the Azure Functions hosted service from what i can tell, which means it's either running in some other Azure service or on some local infrastructure you own (including a laptop or desktop).

To verify if this is the case, try assigning a unique [task hub name](https://docs.microsoft.com/azure/azure-functions/durable/durable-functions-task-hubs) to your function app and seeing if that makes a difference. Doing so will ensure that your app in Azure is isolated from any other app listening on the same storage account.

I haven't looked at any of the other cases yet, but I suspect it may be the same root cause.

**Comment by cgillum** [comment]
@jtoniolo I didn't see any record of the instance id you mentioned. Can you double check the information you provided?

I did see that there are some odd-looking exceptions in our internal telemetry for your app. Here is a JSON parsing exception:

```
Newtonsoft.Json.JsonReaderException: Error parsing boolean value. Path '', line 1, position 1.
   at Newtonsoft.Json.JsonTextReader.ParseTrue()
   at Newtonsoft.Json.JsonTextReader.ParseValue()
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.ReadForType(JsonReader reader, JsonContract contract, Boolean hasConverter)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
   at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
   at Newtonsoft.Json.JsonConvert.DeserializeObject(String value, Type type, JsonSerializerSettings settings)
   at Newtonsoft.Json.JsonConvert.DeserializeObject[T](String value, JsonSerializerSettings settings)
   at DurableTask.AzureStorage.MessageManager.<DeserializeQueueMessageAsync>d__12.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at DurableTask.AzureStorage.Messaging.WorkItemQueue.<GetMessageAsync>d__5.MoveNext()
```

By chance is someone manually writing messages to your work item queue (**bdotasksdev-workitems**)? This is never expected since we control the format of all messages internally. Switching between old and new versions of the extension could potentially cause this as well if there was an accidentally compatibility change that was introduced between 1.7.1 (which is quite old) and 1.8.3.

If possible, I recommend you open a new issue specific to your app and provide an instance ID for an app with the latest extension version.

**Comment by cgillum** [comment]
@graham-saunders it looks like you're running a very old version of the extension (v1.6.2). The latest is v1.8.3. I don't know if that's related to this issue, but there have definitely been some important reliability fixes between these releases. I recommend upgrading as soon as you get a chance.

Also, it looks like your task hub name for that instance is **InterminableJobsHub**, not _InterminableJobsHubInstances_. Can you confirm?

Lastly, and probably most importantly, I see the following error that seems to have impacted processing of your

```
Newtonsoft.Json.JsonSerializationException: Type specified in JSON 'DurableTask.AzureStorage.MessageData' was not resolved. Path '$type', line 1, position 47.
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.ResolveTypeName(JsonReader reader, Type& objectType, JsonContract& contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, String qualifiedTypeName)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.ReadMetadataProperties(JsonReader reader, Type& objectType, JsonContract& contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue, Object& newValue, String& id)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
   at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
   at Newtonsoft.Json.JsonConvert.DeserializeObject(String value, Type type, JsonSerializerSettings settings)
   at Newtonsoft.Json.JsonConvert.DeserializeObject[T](String value, JsonSerializerSettings settings)
   at DurableTask.AzureStorage.MessageManager.<DeserializeQueueMessageAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at DurableTask.AzureStorage.Messaging.ControlQueue.<>c__DisplayClass12_0.<<GetMessagesAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at DurableTask.AzureStorage.Utils.<ParallelForEachAsync>d__2`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at DurableTask.AzureStorage.Messaging.ControlQueue.<GetMessagesAsync>d__12.MoveNext()
```

My guess is that this has something to do with the Azure Functions runtime version as @ConnorMcMahon mentioned. While it may have been released ~24 days ago, your app didn't actually pick it up until 2019-08-09 02:13:46 UTC.

Also, if this is impacting hundreds of your tenants, I suggest opening a support case with Azure to make sure this issue is getting correctly tracked and prioritized.

**Comment by ConnorMcMahon** [comment]
@marcin-synak I didn't see any obvious errors or warnings in our logs for the orchestration instance 2c7ccee859554819bd214ca9eb5bb9d9, and it looks like all of the scheduled tasks were successfully processed.

Can you create a separate issue, since it looks unrelated to the issues of others in this thread. If at all possible, could you provide us with a code sample that causes one of the stuck orchestrations? I want to rule out that there is no non-deterministic code running in the orchestration.

In the meantime, I am closing this issue, as the original issue is resolved, and the other issues in the thread are different.
